torch/csrc/jit/codegen/cuda/lower_vectorize_welford.cpp

```diff
@@ -0,0 +1,567 @@
+    }
+
+    // nvfuser_index_t new_count;
+    // new_count = hoisted_count + count_increment
+
+    auto new_count = defineScalar(index_type);
+
+    auto hoisted_count = hoistCount(wop->outN()->as<kir::TensorIndex>());
+
+    Int* count_increment = nullptr;
+    if (!is_predicated) {
+      count_increment = GpuLower::current()->kernel()->oneVal();
+    } else {
+      // count_increment = (int)pred;
+      count_increment = defineScalar(index_type)->as<Int>();
+      registerInsertBeforeInnerMostLoop(
+          IrBuilder::create<UnaryOp>(UnaryOpType::Cast, count_increment, pred));
+    }
+
+    registerInsertBeforeInnerMostLoop(IrBuilder::create<BinaryOp>(
+        BinaryOpType::Add, new_count, hoisted_count, count_increment));
+
+    // float new_count_float;
+    auto new_count_float = defineScalar(data_type);
+
+    // new_count_float = (float)new_count;
+    registerInsertBeforeInnerMostLoop(IrBuilder::create<UnaryOp>(
+        UnaryOpType::Cast, new_count_float, new_count));
+
+    // float reciprocal;
+    auto reciprocal = defineScalar(data_type);
+
+    auto reciprocal_expr = IrBuilder::create<BinaryOp>(
+        BinaryOpType::Div,
+        reciprocal,
+        GpuLower::current()->kernel()->oneVal(),
+        new_count_float);
+
+    // If not predicated, just set the reciprocal variable
+    // with the reciprocl expr. Otherwise, guard it with an if
+    // statement.
+    if (!is_predicated) {
+      registerInsertBeforeInnerMostLoop(reciprocal_expr);
+    } else {
+      // Initialize reciprocal as 0;
+      registerInsertBeforeInnerMostLoop(IrBuilder::create<UnaryOp>(
+          UnaryOpType::Set,
+          reciprocal,
+          GpuLower::current()->kernel()->zeroVal()));
+
+      // if (pred) reciprocal = 1 / new_count_float;
+      auto reciprocal_ite = IrBuilder::create<kir::IfThenElse>(
+          IrBuilder::create<kir::Predicate>(pred));
+      registerInsertBeforeInnerMostLoop(reciprocal_ite);
+      registerInsertBefore(
+          nullptr, reciprocal_expr, &(reciprocal_ite->thenBody()));
+    }
+
+    auto vectorized_wop = IrBuilder::create<kir::VectorizedWelfordOp>(
+        wop->outputTriplet(),
+        wop->inputTriplet(),
+        wop->initTriplet(),
+        new_count,
+        reciprocal,
+        pred);
+
+    return vectorized_wop;
+  }
+
+  // Declare a scalar variable of type dt and insert its allocation
+  Val* defineScalar(DataType dt) {
+    Val* val = IrBuilder::newScalar(dt);
+
+    auto alloc = IrBuilder::create<kir::Allocate>(
+        val, MemoryType::Local, GpuLower::current()->kernel()->oneVal());
+
+    registerInsertBeforeInnerMostLoop(alloc);
+
+    return val;
+  }
+
+  // Hoist the count TensorIndex out of the innermost loop, assuming
+  // it is invariant within the loop. The loop index can be replaced
+  // with what ever value within the loop range since it is
+  // independent of the loop index.
+  kir::TensorIndex* hoistCount(kir::TensorIndex* out_N) {
```

Comment:
Wonder if it would be convenient to add a utility to detect which loops a `Val` is dependent on so we could assert the invariance this assumes (I know it's really just detected earlier). @zasdfgbnm

Reply:
Yes, that would be useful. We could probably just do a search of a specific `Val` in an index math expression.